### DISCUSSION TITLE
Anyone else experiencing slow tests in Jest related to animations?

### DISCUSSION THREAD

**Discussion by elliothimmelfarb** [chakra-ui/chakra-ui#5449] [Q&A]
# Problem Description
- My organization was quite behind on upgrading Chakra-ui. In the previous version, before Chakra-ui changed its animation library to Framer-Motion, we were able to mock the previous animation library quite simply in Jest and our tests ran quickly without needing to wait for animations to finish.
- Since upgrading to the latest Chakra-ui, using Framer-Motion, we have been struggling with with slow Jest tests. We believe this is because we have not been able to successfully mock Framer-Motion in the same way we mocked the previous animation library.
- We've come across mocking solutions [like this one](https://dev.to/tmikeschu/mocking-framer-motion-v4-19go) but haven't been able to get them working. We think this might be because we use a mono-repo architecture and Jest Manual Mocks needs these dependencies to be in a specific `node_modules` relative to where the tests are being executed.
- We've attempted trying to disable animations all together via global CSS but this also didn't seem to work, likely because Framer-Motion uses more than just CSS for transitions.
- Its very strange to us that we don't seem to be able to find many others with this same problem. It seems that this would be a common problem most people would have experienced after upgrading Chakra-ui.

# Questions
- Has anyone else run into this problem and solved it for their projects?
- Does it seem obvious to anyone that we are perhaps barking up the wrong tree for why our tests are slow after upgrading Chakra-ui to the version that uses Framer-Motion?

Thank you all for reading and answering if you are able to 🙏

**Comment by Yandawl** [comment]
Yes, also experiencing exactly the same problem. Did you happen to find a solution in the end @elliothimmelfarb? 😄

**Reply by elliothimmelfarb** [reply]
I just picked this task up a little over a year after originally posting this and I still can't seem to find a solution to this, even though it's something that many people are asking for.

**Comment by elliothimmelfarb** [comment]
To record efforts, here are a few threads that I found on this topic. I haven't been able to find a solution to this yet.

https://github.com/chakra-ui/chakra-ui/issues/3696

https://github.com/framer/motion/issues/1160

https://github.com/chakra-ui/chakra-ui/issues/3696

**Comment by tmikeschu** [comment]
> * We've come across mocking solutions [like this one](https://dev.to/tmikeschu/mocking-framer-motion-v4-19go) but haven't been able to get them working. We think this might be because we use a mono-repo architecture and Jest Manual Mocks needs these dependencies to be in a specific `node_modules` relative to where the tests are being executed.

😆 😭 I wrote that and while it helped somewhat, we're still experiencing really slow tests with any of the overlay components (menu, popover, modal, etc).

**Reply by elliothimmelfarb** [reply]
Thanks for coming up with that back in the day! Having it implemented was the last time I can remember our team having quick tests. There's got to be a way to turn off animations in a test environment. 

I wonder in which library it will be implemented first, Chakra or Framer.

**Reply by tmikeschu** [reply]
In the meantime, I wrote some stub Menu components that just use basic show/hide context state. I had one menu in particular I wanted to test every button on and the test went from like 60s to 10s. Still slow but tolerable and at least testing the way I want to test.

Here's a gist of the mock we have for chakra-ui/react:
https://gist.github.com/tmikeschu/62bf4f56d9c5e891e44ae5432cf4a38a

It probably has some functionality gaps (I just let the jest errors guide the implementation beyond the basic isOpen/onClose functionality (e.g., `useMenuItem`)), but it should be a good starting point. We plan to do the same for Popover, Tooltip, and Modal.

**Comment by MashaSinkevich** [comment]
https://github.com/chakra-ui/chakra-ui/issues/8033 - I've created a separate improvement ticket as there I have more evidence on low performance in jest + mocking way (not the best one, hopefully we would be able to find a better one).

**Comment by janowskipio** [comment]
Hello everyone! First of all: thanks for the inspiration from the above topic/answers! 🙇 

In my case, we are running 2000+ tests, and animations used for Accordion/Modal/Menu/Popover etc were responsible for ~25%+ tests execution time. 

I agree it would be best to be able to mock the main actor (`framer-motion`), but for now, I went with an easier workaround -> mocking some of the chakra components, which are using `motion` from `framer-motion` ([search results](https://github.com/search?q=repo%3Achakra-ui%2Fchakra-ui+motion+path%3A%2F%5Epackages%5C%2Fcomponents%5C%2F%2F&type=code)).

Below you can find the code I used in the `jest` tests setup file (FYI some props destructuring are specific to components in my project, used `any` type and destructuring just to not pass those into the HTML node to have clean tests output):

```typescript
import { PropsWithChildren } from 'react'

jest.mock('@chakra-ui/react', () => {
  const actualChakra = jest.requireActual('@chakra-ui/react')

  return {
    ...actualChakra,

    usePrefersReducedMotion: () => true,

    Accordion: (props: PropsWithChildren) => (
      <actualChakra.Accordion {...props} reduceMotion />
    ),

    DrawerContent: (props: PropsWithChildren) => (
      <div role="dialog" aria-modal="true" {...props} />
    ),

    MenuList: ({ zIndex, ...rest }: PropsWithChildren & { zIndex: any }) => {
      const { isOpen } = actualChakra.useMenuContext()

      return (
        <div
          role="menu"
          style={{ visibility: isOpen ? 'visible' : 'hidden' }}
          {...rest}
        />
      )
    },

    ModalContent: ({
      width,
      maxWidth,
      ...rest
    }: PropsWithChildren & { width: any; maxWidth: any }) => (
      <div role="dialog" aria-modal="true" {...rest} />
    ),

    PopoverContent: ({
      width,
      maxWidth,
      maxHeight,
      ...rest
    }: PropsWithChildren & { width: any; maxWidth: any; maxHeight: any }) => {
      const { isOpen, getPopoverProps } = actualChakra.usePopoverContext()
      const { role } = getPopoverProps()

      return isOpen ? <section role={role} {...rest} /> : null
    },
  }
})
``` 

You can probably toggle between `isOpen ? ... : null` and `style={{ visibility: isOpen ? 'visible' : 'hidden' }}` implementation, based on what result you expect to have in the tests (rendering (or not) content in the HTML tree even when closed).

Also FYI as you can check above I went with `reduceMotion` for the `Accordion`, but I don't have many tests with components using Accordion, so I'm not able to tell if it makes any difference (so not sure if I shouldn't mock `Collapse` component which is using `motion` instead)

The next step for me is to try to mock `motion` from `framer-motion` - to have one mock for all chakra animated components, maybe I will have time to spend some more time on it this year 👋
